core/src/commonMain/kotlin/com/xebia/functional/xef/llm/Chat.kt

```diff
@@ -37,19 +34,14 @@ interface Chat : LLM {
         streamToStandardOut = true
       )
 
-    val buffer = StringBuilder()
     createChatCompletions(request)
-      .onEach {
-        it.choices.forEach { choice ->
-          val text = choice.delta?.content ?: ""
-          buffer.append(text)
-        }
-      }
-      .onCompletion {
-        val message = assistant(buffer.toString())
+      .map { it.choices.mapNotNull { it.delta?.content }.reduce(String::plus) }
+      .onEach { emit(it) }
+      .fold("", String::plus)
```

Comment:
Is there a better way to achieve the same as the `onEach` and `fold`? Currently, each item is emitted to another outer flow.
This is required because we have to `fold` the flow and get the full generated string in order to add it to the memory. On the other hand, we want to pass each item forward to the caller of the function so that they can do whatever the want to do with it.

Reply:
It kind-of depends on the desired behavior of the resulting `Flow`.
`onEach` will back-pressure the `fold`, so the resulting `Flow` can still back-pressure the `fold` which is desired.

`onEach` is the same of `map { f(it); it }`. 

My concern is that `fold` will take everything in memory, but I guess that's inherit to `MemoryManagement.addMemoriesAfterStream`. Just going by the name it sounds like it puts everything in memory anyway.

Reply:
@Intex32 I explain back-pressure a little bit here. https://youtu.be/n724AdSAkOI
Let me know if this clears it up for you, if not we can discuss it further ☺️

Reply:
@nomisRev I have more questions than before 😂. I especially liked the part where you talked about back pressure.
The pen has some classy dance moves tbh.